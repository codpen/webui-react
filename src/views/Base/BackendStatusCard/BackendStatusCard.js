import React from "react";
import {Button, Card, CardBody, CardHeader} from "reactstrap";
import * as PropTypes from "prop-types";
import ReactDOM from "react-dom";
import RunningJobs from "../RunningJobs";
import {connect} from "react-redux";
import {enableCheckStatus, getStatus} from "../../../actions/statusActions";
import {MODAL_ROOT_ELEMENT, STATUS_REFRESH_TIMEOUT} from "../../../utils/Constants";

/**
 * Functional component Modal which is placed in the element with id "modal-root" in index.html using React.createPortal
 * @returns {{children, implementation, containerInfo, $$typeof, key}}
 * @constructor
 */
function TaskModal() {
    return ReactDOM.createPortal((
        <RunningJobs mode={"modal"}/>

    ), document.getElementById(MODAL_ROOT_ELEMENT));
}

/**
 * Component for display and monitoring of backend rclone status. Auto refresh status in redux store every 5 seconds.
 */
class BackendStatusCard extends React.Component {


    componentWillMount() {

        // Check if the connection to the backend is active
        this.props.getStatus();
        this.refreshInterval = setInterval(() => this.props.getStatus(), STATUS_REFRESH_TIMEOUT);
    }


    componentWillUnmount() {
        // Clear the interval before component is unmounted
        clearInterval(this.refreshInterval);
    }

    /**
     * Enable or disable checking of status request by http request to the backend.
     */
    toggleCheckStatus = () => {
        const {checkStatus, enableCheckStatus} = this.props;
        console.log(checkStatus, enableCheckStatus);
        enableCheckStatus(!checkStatus);
    };

    /**
     * Renders the component with mode.
     * Card: Enables the card mode.
     * Default: Table mode (Grid)
     * @returns {*}
     */
    render() {
        const {isConnected, mode, checkStatus} = this.props;

        const ipAddress = localStorage.getItem('ipAddress');
        const username = localStorage.getItem("username");


        if (mode === "card")
            return (

                <Card
                    className={"text-center " + (isConnected ? "card-accent-info" : "card-accent-warning")}>
                    <CardHeader>
                        rclone status
                    </CardHeader>
                    <CardBody>
                        <StatusText checkStatus={checkStatus} connectivityStatus={isConnected} ipAddress={ipAddress}
                                    userName={username}/>

                    </CardBody>
                </Card>
            );
        else /*Default*/
            return (
                <React.Fragment>
                    <Button type="primary" onClick={this.toggleCheckStatus}
                            className={isConnected ? "bg-info  d-none d-lg-block" : "bg-warning d-none d-lg-block"}> {checkStatus ? isConnected ? "CONNECTED" : "DISCONNECTED" : "DISABLED"}</Button>
                    {/*Show current tasks in the side modal*/}
                    <TaskModal/>
                </React.Fragment>
            );
    }
}

/**
 *
 * @param connectivityStatus    {boolean} Current connectivity status to the backend.
 * @param checkStatus           {boolean} Specify whether to check the status or skip.
 * @param ipAddress             {string} IP Address of the backend
 * @param userName              {string} User name of the currently logged in user.
 * @returns {*}
 * @constructor
 */
function StatusText({connectivityStatus, checkStatus, ipAddress, userName}) {
    if (!checkStatus) {
        return <p>Not monitoring connectivity status. Tap the icon in navbar to start.</p>
    }
    if (connectivityStatus) {
        return (
            <p>The rclone backend is connected and working as expected.<br/>Current IP address is {ipAddress}
                <br/><strong> Username: </strong>{userName}</p>
        );
    } else {
        return (
            <p>Cannot connect to rclone backend. There is a problem with connecting to {ipAddress}.</p>

        )
    }
}

const propTypes = {
    /**
     * Used to specify mode of render : card/ grid.
     */
    mode: PropTypes.string.isRequired,
    isConnected: PropTypes.bool.isRequired,
    checkStatus: PropTypes.bool.isRequired,
    enableCheckStatus: PropTypes.func.isRequired,
    getStatus: PropTypes.func.isRequired
};

const defaultProps = {
    mode: "card",
};


BackendStatusCard.propTypes = propTypes;
BackendStatusCard.defaultProps = defaultProps;

const mapStateToProps = state => ({
    isConnected: state.status.isConnected,
    isDisabled: state.status.isDisabled,
    checkStatus: state.status.checkStatus
});

export default connect(mapStateToProps, {getStatus, enableCheckStatus})(BackendStatusCard);
