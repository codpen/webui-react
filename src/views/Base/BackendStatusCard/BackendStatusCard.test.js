import React from "react";
import {shallow} from "enzyme";
import {testStore} from "../../../../Utils";
import BackendStatusCard from "./BackendStatusCard";
import toJson from "enzyme-to-json";

const setUp = (intialState = {}, props = {}) => {
    const store = testStore(intialState);
    const component = shallow(<BackendStatusCard {...props} store={store}/>);
    return component.childAt(0).dive();
}

describe('Backend Status Card', function () {

    describe('renders showChangeBandwidth: true', function () {
        let wrapper;
        beforeEach(() => {
            const props = {
                showChangeBandwidth: true
            };
            const initialState = {
                status: {
                    isConnected: false,
                    jobs: {},
                    checkStatus: true
                },
            };
            wrapper = setUp(initialState, props)
        });

        it('should render without crashing', function () {
            expect(wrapper).toHaveLength(1)
        });

        it('should match snapshot', function () {
            expect(toJson(wrapper)).toMatchSnapshot()
        });
    });

    describe('renders showChangeBandwidth: false', function () {
        let wrapper;
        beforeEach(() => {
            const props = {
                showChangeBandwidth: false
            };
            const initialState = {
                status: {
                    isConnected: false,
                    jobs: {},
                    checkStatus: true
                },
            };
            wrapper = setUp(initialState, props)
        });

        it('should render without crashing', function () {
            expect(wrapper).toHaveLength(1)
        });

        it('should match snapshot', function () {
            expect(toJson(wrapper)).toMatchSnapshot()
        });
    });



});