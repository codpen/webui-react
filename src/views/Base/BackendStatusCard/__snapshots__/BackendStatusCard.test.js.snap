// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend Status Card renders should match snapshot 1`] = `
<Card
  className="text-center card-accent-warning"
  tag="div"
>
  <CardHeader
    tag="div"
  >
    rclone status
  </CardHeader>
  <CardBody
    tag="div"
  >
    <StatusText
      connectivityStatus={false}
      ipAddress={null}
      userName={null}
    />
  </CardBody>
</Card>
`;
